dotnet/src/Experimental/Agents/Internal/Agent.cs

```diff
@@ -152,6 +174,52 @@ public async Task DeleteThreadAsync(string? id, CancellationToken cancellationTo
         await this._restContext.DeleteThreadModelAsync(id!, cancellationToken).ConfigureAwait(false);
     }
 
+    /// <inheritdoc/>
+    public Task AddFileAsync(string fileId, CancellationToken cancellationToken = default)
+    {
+        if (this._isDeleted)
+        {
+            return Task.CompletedTask;
+        }
+
+        if (this._fileIds.Contains(fileId))
+        {
+            return Task.CompletedTask;
+        }
+
+        try
+        {
+            return this._restContext.AddAssistantFileAsync(this.Id, fileId, cancellationToken);
+        }
+        finally
+        {
+            this._fileIds.Add(fileId);
```

Comment:
Why put this in the "finally"?  This can lead to mismatches from reality if AddAssistantFileAsync() fails, not?

Same with RemoveFileAsync()

Reply:
Good point...I think I was playing with this before I moved to the definitive hashset.  Needs fix.